manager/controllers/app/moduleinstance.go

```diff
@@ -68,87 +83,115 @@ func (m *ModuleManager) GetCopyDestination(item modules.DataInfo, destinationInt
 			S3: &pb.S3DataStore{
 				Bucket:    bucket.Spec.Name,
 				Endpoint:  bucket.Spec.Endpoint,
-				ObjectKey: bucket.Status.AssetPrefixPerDataset[item.AssetID],
+				ObjectKey: bucket.Status.AssetPrefixPerDataset[item.Context.DataSetID],
 			},
 		},
 		Format: string(destinationInterface.DataFormat),
 	}, nil
 }
 
-// SelectModuleInstances selects the necessary read/copy/write modules for the blueprint
-func (m *ModuleManager) SelectModuleInstances(item modules.DataInfo) ([]modules.ModuleInstanceSpec, error) {
-	instances := make([]modules.ModuleInstanceSpec, 0)
+func (m *ModuleManager) selectReadModule(item modules.DataInfo, appContext *app.M4DApplication) (*modules.Selector, error) {
+	// read module is required if the workload exists
+	if appContext.Spec.Selector.WorkloadSelector.Size() == 0 {
+		return nil, nil
+	}
+	m.Log.Info("Select read path for " + item.Context.DataSetID)
+	// select a read module that supports user interface requirements
+	// actions are not checked since they are not necessarily done by the read module
+	readSelector := &modules.Selector{Flow: app.Read,
+		Destination:  &item.Context.Requirements.Interface,
+		Actions:      make([]pb.EnforcementAction, 0),
+		Source:       nil,
+		Dependencies: make([]*app.M4DModule, 0),
+		Module:       nil,
+		Message:      ""}
+	if !readSelector.SelectModule(m.Modules) {
+		m.Log.Info(readSelector.GetError())
+		return nil, errors.New(readSelector.GetError())
+	}
+	return readSelector, nil
+}
 
-	// Write path is not yet implemented
-	var readSelector, copySelector *modules.Selector
-	m.Log.Info("Select read path for " + item.AssetID)
-	// Select a module that supports READ flow, supports actions-on-read, has the required dependency modules (recursively), with API = sink.
-	actionsOnRead := item.Actions[app.Read]
-	if !actionsOnRead.Allowed {
-		return instances, errors.New(actionsOnRead.Message)
+func (m *ModuleManager) selectCopyModule(item modules.DataInfo, appContext *app.M4DApplication, readSelector *modules.Selector) (*modules.Selector, error) {
+	// logic for deciding whether copy module is required
+	var interfaces []*app.InterfaceDetails
+	var copyRequired bool
+	var actionsOnCopy []pb.EnforcementAction
+
+	if readSelector != nil {
+		copyRequired, interfaces, actionsOnCopy = m.getCopyRequirements(item, readSelector)
+	} else if item.Context.Requirements.Copy.Required {
```

Comment:
Please add a comment - User explicitly requested that the data be copied rather than read.

Reply:
User can request copy even if the data is read by the workload (performance reasons, etc.)